src/MongoDB.Driver.Core/Core/Compression/Zstandard/ZstandardNativeWrapper.cs

```diff
@@ -0,0 +1,286 @@
+        private readonly CompressionMode _compressionMode;
+        private bool _operationInitialized;
+        private readonly ulong _recommendedZstreamInputSize;
+        private readonly ulong _recommendedZstreamOutputSize;
+        private IntPtr _zstreamPointer;
+
+        public ZstandardNativeWrapper(CompressionMode compressionMode, int compressionLevel)
+        {
+            _compressionLevel = compressionLevel;
+            _compressionMode = compressionMode;
+
+            switch (_compressionMode)
+            {
+                case CompressionMode.Compress:
+                    // calculate recommended size for input buffer
+                    _recommendedZstreamInputSize = Zstandard64NativeMethods.ZSTD_CStreamInSize();
+                    // calculate recommended size for output buffer. Guarantee to successfully flush at least one complete compressed block
+                    _recommendedZstreamOutputSize = Zstandard64NativeMethods.ZSTD_CStreamOutSize();
+                    _zstreamPointer = Zstandard64NativeMethods.ZSTD_createCStream(); // create resource
+                    break;
+                case CompressionMode.Decompress:
+                    // calculate recommended size for input buffer
+                    _recommendedZstreamInputSize = Zstandard64NativeMethods.ZSTD_DStreamInSize();
+                    // calculate recommended size for output buffer. Guarantee to successfully flush at least one complete block in all circumstances
+                    _recommendedZstreamOutputSize = Zstandard64NativeMethods.ZSTD_DStreamOutSize();
+                    _zstreamPointer = Zstandard64NativeMethods.ZSTD_createDStream(); // create resource
+                    break;
+            }
+        }
+
+        public int RecommendedInputSize => (int)_recommendedZstreamInputSize;
+        public int RecommendedOutputSize => (int)_recommendedZstreamOutputSize;
+
+        // public methods
+        public void Compress(
+            OperationContext operationContext,
+            int inputCompressedSize,
+            int inputUncompressedSize,
+            out int compressedBytesProcessed,
+            out int uncompressedBytesProcessed)
+        {
+            Ensure.IsNotNull(operationContext, nameof(operationContext));
+            Ensure.That(_compressionMode == CompressionMode.Compress, nameof(_compressionMode));
+
+            InitializeIfNotAlreadyInitialized();
+
+            // compressed data
+            var outputNativeBuffer = CreateNativeBuffer(
+                operationContext.CompressedPinnedBufferWalker, // operation result
+                (ulong)inputCompressedSize);
+
+            // uncompressed data
+            var inputNativeBuffer = CreateNativeBuffer(
+                operationContext.UncompressedPinnedBufferWalker,
+                (ulong)inputUncompressedSize);
+
+            // compress inputNativeBuffer to outputNativeBuffer
+            Zstandard64NativeMethods.ZSTD_compressStream(
```

Comment:
`ZSTD_compressStream` returns a value that we are ignoring.

Some of those values are error codes. Should we check the return value for errors? I think so.

Reply:
I think no, we check error codes on the NativeMethods level. I can remove returning value for that level if you want.

Reply:
OK... to me it looked like we were ignoring the error codes.

This is confusing API design.

Why don't you explicitly ignore the return value:

    var ignoredWhatever = Zstandard64NativeMethods.ZSTD_compressStream(...);

so it will be obvious to readers that you are ignoring the return value on purpose.